app/src/main/java/org/olpc/xorobothash/MainActivity.java

```diff
@@ -0,0 +1,157 @@
+        Log.i(TAG, "In updateView()");
+        boolean robotVisible = false;
+        File file = new File(robotImagePath());
+        // Check if we have ever successfully downloaded the XO's robot image
+        if (file.exists()) {
+            Log.i(TAG, "Robohash image file exists.");
+            robotVisible = showRobot(robotImagePath());
+        } else if (!isConnected) {
+            Log.i(TAG, "Robohash image file does not exist, network is not connected.");
+            showDisconnected();
+        }
+        return robotVisible;
+    }
+
+    private boolean showRobot(String imagePath) {
+        Log.i(TAG, "In showRobot()");
+        boolean robotDrawn;
+        Bitmap bitmap = BitmapFactory.decodeFile(imagePath);
+        robotImageView.setImageBitmap(bitmap);
+        // Check if the robot image is visible
+        robotDrawn = robotImageView.getDrawable() == null ? false : true;
+        return robotDrawn;
+    }
+
+    private void showDisconnected() {
+        //Hide the robot text
+        TextView robotText = (TextView) findViewById(R.id.robot_text_message);
+        robotText.setVisibility(View.GONE);
+        //Hide the robot image
+        ImageView robotImage = (ImageView) findViewById(R.id.robot_hash_image);
+        robotImage.setVisibility(View.GONE);
+        //Show the disconnected error message
+        TextView errorMessage = (TextView) findViewById(R.id.error_message);
+        errorMessage.setText(getString(R.string.no_network_connection));
+    }
+
+    private String robotImagePath() {
+        return this.getFilesDir() + File.pathSeparator + "this-xos-robot.png";
+    }
+
+    private String uniqueRoboHashURL() {
+        //TODO test on a real android device
+        String serialNumber = android.os.Build.SERIAL;  // if SERIAL doesn't work, try Settings.Secure.ANDROID_ID; ?
+        return "http://robohash.org/" + serialNumber + ".png?size=500x500";
+    }
+
+    private boolean checkNetworkConnectivity() {
+        boolean connected;
+        ConnectivityManager connMgr = (ConnectivityManager)
+                getSystemService(Context.CONNECTIVITY_SERVICE);
+        NetworkInfo networkInfo = connMgr.getActiveNetworkInfo();
+        if (networkInfo != null && networkInfo.isConnected()) {
+            Log.i(TAG, "Network is connected.");
+            connected = true;
+        } else {
+            Log.e(TAG, "Network is NOT connected.");
+            connected = false;
+        }
+        return connected;
+    }
+
+    private class DownloadRobotImage extends AsyncTask<String, Void, Boolean> {
+
+        @Override
+        protected void onPreExecute() {
+            // TODO Display a spinner or loading message
+        }
+
+        @Override
+        protected Boolean doInBackground(String... params) {
+            Log.i(TAG, "In DownloadRobotImage() doInBackground()");
+            HttpURLConnection connection = null;
+            Boolean downloadSuccess = true;
```

Comment:
Do this the other way round: set it `false` to start and `true` only after download finishes.